File: advance_erp/public/js/quotation.js
frappe.ui.form.on("Quotation", {
    refresh(frm) {
        if (!frm.is_new()) {
            frm.add_custom_button(__('View Last Prices'), function() {
                open_last_prices_dialog(frm);
            });
            frm.add_custom_button("Suggest Price", function() {
                frappe.call({
                    method: "advance_erp.advance_erp.pricing.quotation.get_suggested_prices",
                    args: { quotation_name: frm.doc.name },
                    callback: function(r) {
                        if (r.message && r.message.length) {
                            let items = r.message;
                            let html = `<table class="table table-bordered" style="width:100%; table-layout: fixed;">
                                <thead>
                                    <tr>
                                        <th style="width:20%; background-color: bisque;">Item Code</th>
                                        <th style="width:25%; background-color: bisque;">Item Name</th>
                                        <th style="width:20%; background-color: bisque;">Current Price</th>
                                        <th style="width:20%; background-color: bisque;">Suggested Price</th>
                                        <th style="width:15%; background-color: bisque;">Apply</th>
                                    </tr>
                                </thead>
                                <tbody>`;
                            items.forEach((item, i) => {
                                html += `<tr>
                                    <td style="background: #FFFDE1;">${item.item_code}</td>
                                    <td style="background: #FFFDE1;">${item.item_name}</td>
                                    <td style="background: #FFFDE1;">${item.current_price}</td>
                                    <td style="background: #FFFDE1;">
                                        <input type="text"
                                            class="suggested-price"
                                            data-index="${i}"
                                            value="${item.suggested_price}"
                                            style="width:100%; box-sizing:border-box;"
                                            oninput="this.value = this.value
                                                .replace(/[^0-9.]/g, '')
                                                .replace(/(\..*)\./g, '$1');">
                                    </td>
                                    <td style="text-align: center; background: #FFFDE1;">
                                        <input type="checkbox" class="keep-suggested" data-index="${i}" checked>
                                    </td>
                                </tr>`;
                            });
                            html += `</tbody></table>`;
                            let d = new frappe.ui.Dialog({
                                title: "Review Suggested Prices",
                                fields: [
                                    {
                                        fieldtype: "HTML",
                                        fieldname: "items_html",
                                        options: html
                                    }
                                ],
                                primary_action_label: "Apply Suggested Price",
                                primary_action: function() {
                                    let selected_items = items.map((item, i) => {
                                        let checkbox = d.$wrapper.find(`.keep-suggested[data-index="${i}"]`)[0];
                                        let checked = checkbox ? checkbox.checked : false;
                                        let suggested_price_input = d.$wrapper.find(`.suggested-price[data-index="${i}"]`);
                                        let suggested_price = suggested_price_input ? parseFloat(suggested_price_input.val()) : item.suggested_price;
                                        if (suggested_price < 0 || isNaN(suggested_price)) {
                                            suggested_price = 0;
                                        }
                                        return {
                                            item_code: item.item_code,
                                            keep_suggested: checked,
                                            suggested_price: suggested_price
                                        };
                                    });
                                    frappe.call({
                                        method: "advance_erp.advance_erp.pricing.quotation.apply_suggested_prices",
                                        args: {
                                            quotation_name: frm.doc.name,
                                            items: selected_items
                                        },
                                        callback: function() {
                                            frm.reload_doc();
                                            d.hide();
                                        }
                                    });
                                }
                            });
                            d.show();
                        }
                    }
                });
            });
        }
        if (frm.doc.docstatus === 1) {
            frm.remove_custom_button('View Last Prices');
        }
    }
});


function open_last_prices_dialog(frm) {
    let dialog = new frappe.ui.Dialog({
        title: 'View Last Prices',
        fields: [
            {
                fieldname: 'customer_filter_type',
                label: 'Filter Type',
                fieldtype: 'Select',
                options: ['Current Customer', 'All Customers', 'Customer Group'],
                default: 'Current Customer',
                reqd: 1
            },
            { fieldtype: 'Column Break' },
            {
                fieldname: 'customer_group',
                label: 'Customer Group',
                fieldtype: 'Link',
                options: 'Customer Group',
                depends_on: "eval:doc.customer_filter_type=='Customer Group'"
            },
            { fieldtype: 'Column Break' },
            {
                fieldname: 'item_codes',
                label: 'Item Codes',
                fieldtype: 'MultiSelectList',
                get_data: function (txt) {
                    let items = (frm.doc.items || []).map(row => row.item_code);
                    return items
                        .filter(i => !txt || i.toLowerCase().includes(txt.toLowerCase()))
                        .map(i => ({
                            value: i,
                            description: i
                        }));
                }
            },
            { fieldtype: 'Column Break' },
            {
                fieldname: 'last_n_prices',
                label: 'Number of Last Prices',
                fieldtype: 'Int',
                default: 3
            }
        ],
        size: 'extra-large'
    });
    dialog.show();
    dialog.$wrapper.css({
        width: '1100px',
        height: '80vh',
        marginLeft: '5%',
        marginTop: '5%'
    });
    let $btn_row = $(`
        <div style="display:flex; justify-content:flex-end; margin-bottom:15px;">
            <button class="btn btn-primary">Fetch Prices</button>
        </div>
    `);
    dialog.$body.append($btn_row);
    $btn_row.find('button').on('click', () => {
        let filters = dialog.get_values();
        fetch_last_prices(filters, dialog, frm);
    });
}

function fetch_last_prices(filters, dialog, frm) {
    let customer = null;
    let all_customers = false;
    let customer_group = null;
    if (filters.customer_filter_type === 'Current Customer') {
        customer = frm.doc.customer;
    } 
    else if (filters.customer_filter_type === 'All Customers') {
        all_customers = true;
    } 
    else if (filters.customer_filter_type === 'Customer Group') {
        customer_group = filters.customer_group;
    }
    let quotation_items = (frm.doc.items || []).map(row => row.item_code);
    let item_codes = quotation_items;
    if (Array.isArray(filters.item_codes) && filters.item_codes.length > 0) {
        item_codes = filters.item_codes;
    }
    frappe.call({
        method: 'advance_erp.advance_erp.pricing.quotation.get_last_prices',
        args: {
            customer,
            all_customers,
            customer_group,
            item_codes,
            last_n_prices: filters.last_n_prices
        },
        callback: function (r) {
            if (r.message && r.message.length) {
                show_last_prices_table(r.message, dialog, frm);
            } else {
                frappe.msgprint(__('No prices found'));
            }
        }
    });
}


function show_last_prices_table(prices, dialog, frm) {
    if (dialog.$table_wrapper) dialog.$table_wrapper.remove();
    dialog.$table_wrapper = $('<div style="max-height:280px; overflow-y:auto; border:1px solid black;"></div>').appendTo(dialog.body);
    const colors = ['skyblue', 'pink'];
    const colorMap = {};
    let colorIndex = 0;
    let table_rows = prices.map((row, i) => {
        if (!colorMap[row.item_code]) {
            colorMap[row.item_code] = colors[colorIndex % 2];
            colorIndex++;
        }
        const bgColor = colorMap[row.item_code];
        return `
            <tr data-index="${i}" style="background:${bgColor}">
                <td>${row.customer}</td>
                <td>${row.item_code}</td>
                <td>${row.item_name}</td>
                <td>${row.sales_invoice}</td>
                <td>${row.posting_date}</td>
                <td>${row.rate}</td>
                <td><input type="checkbox" class="choose-checkbox" data-index="${i}"></td>
            </tr>
        `;
    }).join('');
    let table_html = `
        <table class="table table-bordered" style="width:100%; border-collapse:collapse; margin: 0;">
            <thead style="background-color: bisque; position: sticky; top: 0; z-index: 10;">
                <tr>
                    <th>Customer</th>
                    <th>Item Code</th>
                    <th>Item Name</th>
                    <th>Sales Invoice</th>
                    <th>Posting Date</th>
                    <th>Rate</th>
                    <th>Choose</th>
                </tr>
            </thead>
            <tbody>
                ${table_rows}
            </tbody>
        </table>
    `;
    dialog.$table_wrapper.html(table_html);
    dialog.$table_wrapper.on('change', '.choose-checkbox', function () {
        let index = $(this).data('index');
        let item_code = prices[index].item_code;
        if (this.checked) {
            prices.forEach((r, i) => {
                if (i != index && r.item_code == item_code) {
                    r.choose = false;
                    dialog.$table_wrapper.find(`input.choose-checkbox[data-index="${i}"]`).prop('checked', false);
                }
            });
        }
        prices[index].choose = this.checked;
    });
    if (dialog.$apply_btn) {
        dialog.$apply_btn.remove();
    }
    if (dialog.$button_wrapper) {
        dialog.$button_wrapper.remove();
    }
    dialog.$button_wrapper = $(`
        <div style="
            display: flex;
            justify-content: flex-end;
            margin-top: 10px;
        "></div>
    `).appendTo(dialog.body);

    dialog.$apply_btn = $(`<button class="btn btn-primary">Apply Prices</button>`)
        .appendTo(dialog.$button_wrapper)
        .click(() => {
            apply_selected_prices(frm, prices);
            dialog.hide();
        });
}

function apply_selected_prices(frm, table_data) {
    let updated = false;
    table_data.forEach(row => {
        if (row.choose) {
            let quotation_item = frm.doc.items.find(
                i => i.item_code === row.item_code
            );
            if (quotation_item) {
                frappe.model.set_value(
                    quotation_item.doctype,
                    quotation_item.name,
                    "rate",
                    row.rate
                );
                updated = true;
            }
        }
    });
    if (updated) {
        frm.trigger("calculate_taxes_and_totals");
        frm.dirty();
        frm.refresh_field("items");
        frappe.msgprint(__('Prices Applied.'));
    }
}

File: advance_erp/advance_erp/pricing/quotation.py
import frappe
from frappe.utils import add_days, today
import json

@frappe.whitelist()
def get_last_prices(item_codes, customer=None, all_customers=False, customer_group=None, last_n_prices=5):
    print(item_codes)
    if isinstance(item_codes, str):
        item_codes = json.loads(item_codes)
    customer_conditions = []
    if not all_customers:
        if customer:
            customer_conditions.append(f"si.customer = '{customer}'")
        elif customer_group:
            customers = frappe.get_all('Customer', filters={'customer_group': customer_group}, pluck='name')
            if customers:
                customers_str = ",".join([f"'{c}'" for c in customers])
                customer_conditions.append(f"si.customer IN ({customers_str})")
            else:
                return []
    customer_where = " AND ".join(customer_conditions)
    if customer_where:
        customer_where = " AND " + customer_where
    all_prices = []
    for item in item_codes:
        query = f"""
            SELECT
                si.name as sales_invoice,
                si.customer,
                sii.item_code,
                sii.item_name,
                sii.rate,
                si.posting_date
            FROM `tabSales Invoice Item` sii
            JOIN `tabSales Invoice` si ON sii.parent = si.name
            WHERE sii.item_code = '{item}' {customer_where}
            ORDER BY si.posting_date DESC
            LIMIT {last_n_prices}
        """
        results = frappe.db.sql(query, as_dict=True)
        for r in results:
            r['choose'] = 0
        all_prices.extend(results)
    return all_prices

@frappe.whitelist()
def get_suggested_prices(quotation_name):
    doc = frappe.get_doc("Quotation", quotation_name)
    period_year = 1
    interval_days = 30
    threshold_percent = 20
    adjust_percent = 10
    start_date = add_days(today(), -365*period_year)
    items_list = []
    for item in doc.items:
        total_sales = frappe.get_all(
            "Sales Order Item",
            filters={
                "item_code": item.item_code,
                "docstatus": 1,
                "creation": [">=", start_date]
            },
            fields=["sum(qty*rate) as total_value"]
        )
        total_value = total_sales[0].total_value or 0
        intervals = (365*period_year) / interval_days
        avg_sales_per_interval = total_value / intervals
        recent_start = add_days(today(), -interval_days)
        recent_sales = frappe.get_all(
            "Sales Order Item",
            filters={
                "item_code": item.item_code,
                "docstatus": 1,
                "creation": [">=", recent_start]
            },
            fields=["sum(qty*rate) as recent_value"]
        )
        recent_value = recent_sales[0].recent_value or 0
        if avg_sales_per_interval == 0:
            velocity = 0
        else:
            velocity = (recent_value - avg_sales_per_interval) / avg_sales_per_interval * 100
        if velocity > threshold_percent:
            suggested_price = item.rate * (1 + adjust_percent / 100)
        elif velocity < -threshold_percent:
            suggested_price = item.rate * (1 - adjust_percent / 100)
        else:
            suggested_price = item.rate
        suggested_price = round(suggested_price, 2)
        try:
            log = frappe.get_doc({
                "doctype": "Suggested Price Log",
                "quotation": quotation_name,
                "quotation_item": item.name,
                "current_price": item.rate,
                "base_price": item.rate,
                "demand_velocity_percent": round(velocity,2),
                "suggested_price": suggested_price,
                "applied": 0
            })
            log.insert(ignore_permissions=True)
        except Exception as e:
            frappe.log_error(f"Error saving Suggested Price Log: {e}")
        items_list.append({
            "item_code": item.item_code,
            "item_name": item.item_name,
            "current_price": item.rate,
            "suggested_price": suggested_price,
            "keep_suggested": 1
        })
    return items_list


@frappe.whitelist()
def apply_suggested_prices(quotation_name, items):
    if isinstance(items, str):
        items = json.loads(items)
    doc = frappe.get_doc("Quotation", quotation_name)
    for row in doc.items:
        for item in items:
            if isinstance(item, dict) and row.item_code == item.get("item_code"):
                if item.get("keep_suggested"):
                    row.rate = round(item.get("suggested_price") or row.rate, 2)
                    frappe.db.set_value("Suggested Price Log",
                                        {"quotation": quotation_name, "quotation_item": row.name},
                                        "applied", 1)
                else:
                    log_name = frappe.db.get_value("Suggested Price Log",
                                        {"quotation": quotation_name, "quotation_item": row.name})
                    if log_name:
                        frappe.delete_doc("Suggested Price Log", log_name, ignore_permissions=True)
    doc.save()
    return True
